### ISSUE TITLE
Did the API change?

### ISSUE THREAD

**Issue by kurt-rhee** [NREL/developer.nrel.gov#146]
Hello, were there recent changes to the API?  I haven't changed my code at all, but am now receiving error message:

urllib.error.HTTPError: HTTP Error 404: Not Found

Thank you for your help.

**Comment by GUI** [comment]
@kurt-rhee: Can you tell us which specific API you're referring to, and the URL you're trying access? Thanks!

**Comment by kurt-rhee** [comment]
Hello here is my code with api key hidden:

    # Declare all variables as strings. Spaces must be replaced with '+',
    # i.e., change 'John Smith' to 'John+Smith'.
    # Define the lat, long of the location and the year
    years = [1998, 1999, 2000,
             2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010,
             2011, 2012, 2013, 2014, 2015]

    # You must request an NSRDB api key from the link above
    api_key = 'xxxxxxxxxxxxxxxxx'

    # Your full name, use '+' instead of spaces.
    your_name = 'Kurt+Rhee'

    # Your email address
    email = 'xxxxx@edf-re.com'

    dflist = []
    print('Attempting PSMv3 Algorithm:')
    for year in years:
        result = 'None'
        iteration = 0
        while (result == 'None') or (iteration <= 50):
            iteration += 1
            try:

                url = ('http://developer.nrel.gov/api/solar/nsrdb_psm3_download.csv?'
                       'api_key={api_key}&'
                       'full_name={your_name}'
                       '&email={email}'
                       '&affiliation=EDF'
                       '&reason=Prospecting'
                       '&mailing_list=true'
                       '&wkt=POINT({longitude}+{latitude})'
                       '&names={year}'
                       '&attributes=dhi,dni,ghi,clearsky_dhi,clearsky_dni,clearsky_ghi,cloud_type,dew_point,air_temperature,surface_pressure,relative_humidity,total_precipitable_water,wind_direction,wind_speed,fill_flag,surface_albedo'
                       '&leap_day=false'
                       '&utc=false'
                       '&interval=60'.format(api_key=api_key,
                                             your_name=your_name,
                                             email=email,
                                             longitude=longitude,
                                             latitude=latitude,
                                             year=year))
                df = pd.read_csv(url, header=2)


                url = ('http://developer.nrel.gov/api/solar/nsrdb_psm3_download.csv?'
                       'api_key={api_key}&'
                       'full_name={your_name}'
                       '&email={email}'
                       '&affiliation=EDF'
                       '&reason=Prospecting'
                       '&mailing_list=true'
                       '&wkt=POINT({longitude}+{latitude})'
                       '&names={year}'
                       '&attributes=dhi,dni,ghi,clearsky_dhi,clearsky_dni,clearsky_ghi,cloud_type,dew_point,air_temperature,surface_pressure,relative_humidity,total_precipitable_water,wind_direction,wind_speed,fill_flag,surface_albedo'
                       '&leap_day=false'
                       '&utc=false'
                       '&interval=60'.format(api_key=api_key,
                                             your_name=your_name,
                                             email=email,
                                             longitude=longitude,
                                             latitude=latitude,
                                             year=year))
                df = pd.read_csv(url, header=2)

                # get new pandas timeseries index
                df['datetime'] = pd.to_datetime(df[['Year', 'Month', 'Day', 'Hour']])
                df.index = df.datetime

                dflist.append(df)
                result = 'Found'
                
                print('{0} Success'.format(year))

            except Exception as e:
                print('\n')
                print('{0} Failure'.format(year))
                print(e)
                pass

    df = pd.concat(dflist, axis=0)
    df['GHI'] = df['GHI'] / 1000
    df['DHI'] = df['DHI'] / 1000

    # Resample by both sum and average
    df_y = df.resample('Y').sum()

    # Get Time Series Outputs
    ghi = df_y['GHI'].mean()
    variability = float(df_y['GHI'].std() / ghi * 100)
    n_years = len(df_y['GHI'])

**Comment by GUI** [comment]
@PjEdwards: It looks like this is related to the `/api/solar/nsrdb_psm3_download.csv` API. Any recent changes there that would be leading to 404s?

**Comment by kurt-rhee** [comment]
Here is the example from the nrel website with my API Key
https://developer.nrel.gov/docs/solar/nsrdb/psm3_data_download/#examples

```
    url = f'http://developer.nrel.gov/api/solar/nsrdb_psm3_download.csv?api_key={api_key}&full_name=Sample+User&email' \
          f'=sample@email.com&affiliation=Test+Organization&reason=Example&mailing_list=true&wkt=POINT(' \
          f'-108.5449+40.5137)&names=2014&attributes=dhi,dni,ghi,clearsky_dhi,clearsky_dni,clearsky_ghi,cloud_type,' \
          f'dew_point,air_temperature,surface_pressure,relative_humidity,solar_zenith_angle,total_precipitable_water,' \
          f'wind_direction,wind_speed,fill_flag&leap_day=false&utc=false&interval=30'
    df = pd.read_csv(url, header=2)
```

Traceback:

`Traceback (most recent call last):
  File "C:\Users\Simon.Rhee\AppData\Local\Continuum\miniconda3\envs\epe2.0\lib\site-packages\IPython\core\interactiveshell.py", line 3291, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-8-674bc4710d1d>", line 6, in <module>
    df = pd.read_csv(url, header=2)
  File "C:\Users\Simon.Rhee\AppData\Local\Continuum\miniconda3\envs\epe2.0\lib\site-packages\pandas\io\parsers.py", line 702, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "C:\Users\Simon.Rhee\AppData\Local\Continuum\miniconda3\envs\epe2.0\lib\site-packages\pandas\io\parsers.py", line 413, in _read
    filepath_or_buffer, encoding, compression)
  File "C:\Users\Simon.Rhee\AppData\Local\Continuum\miniconda3\envs\epe2.0\lib\site-packages\pandas\io\common.py", line 202, in get_filepath_or_buffer
    req = _urlopen(filepath_or_buffer)
  File "C:\Users\Simon.Rhee\AppData\Local\Continuum\miniconda3\envs\epe2.0\lib\urllib\request.py", line 223, in urlopen
    return opener.open(url, data, timeout)
  File "C:\Users\Simon.Rhee\AppData\Local\Continuum\miniconda3\envs\epe2.0\lib\urllib\request.py", line 532, in open
    response = meth(req, response)
  File "C:\Users\Simon.Rhee\AppData\Local\Continuum\miniconda3\envs\epe2.0\lib\urllib\request.py", line 642, in http_response
    'http', request, response, code, msg, hdrs)
  File "C:\Users\Simon.Rhee\AppData\Local\Continuum\miniconda3\envs\epe2.0\lib\urllib\request.py", line 570, in error
    return self._call_chain(*args)
  File "C:\Users\Simon.Rhee\AppData\Local\Continuum\miniconda3\envs\epe2.0\lib\urllib\request.py", line 504, in _call_chain
    result = func(*args)
  File "C:\Users\Simon.Rhee\AppData\Local\Continuum\miniconda3\envs\epe2.0\lib\urllib\request.py", line 650, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found`

**Comment by kanderso-nrel** [comment]
@GUI This appears to be because the API is blocking the default user-agent string for urllib, which pandas uses under the hood for `pd.read_csv`.  

For instance:

```
import urllib
key = ...
url = f"http://developer.nrel.gov/api/solar/nsrdb_psm3_download.csv?api_key={key}&full_name=Sample+User&email=sample@email.com&affiliation=Test+Organization&reason=Example&mailing_list=true&wkt=POINT(-108.5449+40.5137)&names=2014&attributes=dhi,dni,ghi,clearsky_dhi,clearsky_dni,clearsky_ghi,cloud_type,dew_point,air_temperature,surface_pressure,relative_humidity,solar_zenith_angle,total_precipitable_water,wind_direction,wind_speed,fill_flag&leap_day=false&utc=false&interval=30"

# raises 404
req = urllib.request.Request(url, headers={'User-Agent': "Python-urllib/3.7"})
response = urllib.request.urlopen(req)

# works
req = urllib.request.Request(url, headers={'User-Agent': "some other string"})
response = urllib.request.urlopen(req)
```

**Comment by kanderso-nrel** [comment]
@kurt-rhee, try this workaround for your code.  The `requests` package uses a different default user-agent string that isn't getting rejected. 

```
from io import StringIO
import requests
response = requests.get(url)
buffer = StringIO(response.content.decode('utf-8'))
df = pd.read_csv(buffer, header=2)
```

**Comment by kurt-rhee** [comment]
@kanderso-nrel that works!  Thank you

**Comment by GUI** [comment]
@kurt-rhee, @kanderso-nrel: Thanks for reporting this and figuring out a workaround, but definitely sorry about the issue. It looks like there was a security layer at NREL that inadvertently started blocking several API user agents. But this should not have been happening. I believe this is now fixed so the `Python-urllib/3.7` and other API library user agents aren't being blocked any longer. But please let us know if you see any issues.
